src/GafferScene/Instancer.cpp

```diff
@@ -323,7 +337,103 @@ class Instancer::EngineData : public Data
 			}
 		}
 
+		void initPrototypes( PrototypeMode &mode, const std::string &index, const std::string &rootsVariable, const StringVectorData *rootsList )
+		{
+			const std::vector<std::string> *rootStrings = nullptr;
+
+			switch( mode )
+			{
+				case IndexedRootsList :
+				{
+					if( const auto *indices = m_primitive->variableData<IntVectorData>( index ) )
+					{
+						m_indices = &indices->readable();
+						if( m_indices->size() != numPoints() )
+						{
+							throw IECore::Exception( "prototypeIndex primitive variable has incorrect size" );
+						}
+					}
+
+					rootStrings = &rootsList->readable();
+
+					break;
+				}
+				case IndexedRootsVariable :
+				{
+					if( const auto *indices = m_primitive->variableData<IntVectorData>( index ) )
+					{
+						m_indices = &indices->readable();
+						if( m_indices->size() != numPoints() )
+						{
+							throw IECore::Exception( "prototypeIndex primitive variable has incorrect size" );
+						}
+					}
+
+					const auto *roots = m_primitive->variableData<StringVectorData>( rootsVariable, PrimitiveVariable::Constant );
+					if( !roots )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must be Constant StringVectorData when using IndexedRootsVariable mode" );
+					}
+
+					rootStrings = &roots->readable();
+					if( rootStrings->empty() )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must specify at least one root location" );
+					}
+
+					break;
+				}
+				case RootPerVertex :
+				{
+					const auto view = m_primitive->variableIndexedView<StringVectorData>( rootsVariable, PrimitiveVariable::Vertex );
+					if( !view )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must be Vertex StringVectorData when using RootPerVertex mode" );
+					}
+
+					m_indices = view->indices();
+					rootStrings = &view->data();
+					if( rootStrings->empty() )
+					{
+						throw IECore::Exception( "prototypeRoots primitive variable must specify at least one root location" );
+					}
+
+					break;
+				}
+			}
+
+			m_roots = new ObjectVector;
+			auto &roots = m_roots->members();
+			roots.reserve( rootStrings->size() );
+
+			std::vector<ConstInternedStringVectorDataPtr> inputNames;
+			inputNames.reserve( rootStrings->size() );
+
+			size_t i = 0;
+			ScenePlug::ScenePath path;
+			for( const auto &root : *rootStrings )
+			{
+				ScenePlug::stringToPath( root, path );
+				roots.emplace_back( new InternedStringVectorData( path ) );
+
+				InternedStringVectorDataPtr name = new InternedStringVectorData;
+				if( path.empty() )
+				{
+					name->writable().emplace_back( std::to_string( i++ ) );
+				}
+				else
+				{
+					name->writable().emplace_back( path.back() );
+				}
+				inputNames.emplace_back( name );
+			}
+
+			m_names = new Private::ChildNamesMap( inputNames );
```

Comment:
I hadn't considered that we could use `ChildNamesMap` here - nice!

Reply:
Yeah, it was quite handy, though I'd wished for a constructor that didn't require a vector-of-vectors.